pkg/sosreport/index.js

```diff
@@ -0,0 +1,180 @@
+                        payload: "fsread1",
+                        binary: "raw",
+                        path: archive,
+                        superuser: true,
+                        external: {
+                            "content-disposition": 'attachement; filename="' + basename + '"',
+                            "content-type": "application/octet-stream"
+                        }
+                    }));
+                    sos_archive_url = "/cockpit/channel/" + cockpit.transport.csrf_token + '?' + query;
+                    $("#sos-progress, #sos-error").hide();
+                    $("#sos-alert, #sos-download").show();
+                    $("#sos-cancel").text(_("Close"));
+                } else {
+                    sos_error(_("No archive has been created."), output);
+                }
+                sos_task = null;
+            }
+        });
+        task.fail(function (error) {
+            if (sos_task == task) {
+                sos_error(error.toString(), output);
+                sos_task = null;
+            }
+        });
+    }
+
+    function sos_cancel() {
+        if (sos_task) {
+            if (sos_pid) {
+                /* HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1278784
+                 * SIGTERM only aborts the current plugin, but SIGINT will stop the whole report
+                 */
+                cockpit.spawn([ "pkill", "-SIGINT", "-P", sos_pid ], { superuser: true, err: "message" }).
+                    fail(function (error) {
+                        console.log("failed to kill sosreport", error);
+                    });
+            }
+            sos_task.close();
+            sos_task = null;
+        }
+        if (sos_archive_files.length > 0) {
+            cockpit.spawn([ "rm" ].concat(sos_archive_files), { superuser: true, err: "message" }).
+                fail(function (error) {
+                    console.log("failed to remove", sos_archive_files, error);
+                });
+        }
+        sos_archive_url = null;
+        sos_archive_files =  [ ];
+        $("#sos").modal('hide');
+    }
+
+    function sos_download() {
+        // We download via a hidden iframe to get better control over
+        // the error cases.
+        var iframe = $('<iframe>').attr('src', sos_archive_url);
+        iframe.on('load', function (event) {
+            var title = iframe.get(0).contentDocument.title;
+            if (title)
+                sos_error(title);
+        });
+        $('body').append(iframe);
+        iframe.hide();
+    }
+
+    function init() {
+        $(function () {
+            $("#sos").on("show.bs.modal", sos_init);
+            $("#sos-cancel").on("click", sos_cancel);
+            $('#sos-download button').on('click', sos_download);
+
+            // Establish a transport to the shell so that the tests
+            // know that we are ready to go.
+            //
+            cockpit.transport.wait(function () { });
```

Comment:
The csrf_token above won't be valid until this succeeds. We should probably gate the entire init on this function, no?